### PR TITLE
Use `critical_section` for `Peripherals::take`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- `cortex_m` https://github.com/rust-embedded/cortex-m/pull/447
- `msp430` https://github.com/rust-embedded/msp430/issues/13
- `riscv` https://github.com/rust-embedded/riscv/pull/110
-  `xtensa_lx` https://github.com/esp-rs/xtensa-lx/issues/20, https://github.com/esp-rs/esp-hal/pull/151
- `mips_mcu` https://github.com/kiffie/pic32-rs/issues/5

### CODE REVIEW

File: src/generate/device.rs [hunk 1]
```diff
@@ -300,48 +300,42 @@ pub fn render(d: &Device, config: &Config, device_x: &mut String) -> Result<Toke
...
+        /// All the peripherals.
         #[allow(non_snake_case)]
         pub struct Peripherals {
             #fields
         }
 
         impl Peripherals {
-            #take
+            /// Returns all the peripherals *once*.
+            #[cfg(feature = "critical-section")]
```
Comment (Emilgardis) [inline src/generate/device.rs:318]: This means a new dependency, the crate documentation should be adjusted to account for this.


also, could https://github.com/rust-embedded/svd2rust/blob/a39a61d8ac8928a59d8a2b855c3303a766fd0517/ci/svd2rust-regress/src/svd_test.rs#L8 be updated to include `critical-section`

**Review by Emilgardis — DISMISSED** [review]
lgtm, this is great :D

do we need another reviewer maybe?

**Comment by adamgreig** [comment]
I'm a bit concerned about what this means for platforms that don't have `critical-section` support yet. I know msp430 is working on it but having some codegen size issues, and I don't know anything about `mips_mcu`. It might be worth waiting a little longer for those architectures to get support before releasing svd2rust that uses it?

Having it as an optional dependency/feature seems useful though and does mean those platforms don't need immediate critical-section support, even end-users could add simple interrupt-based implementations to use `take()` if needed.

**Comment by reitermarkus** [comment]
> It might be worth waiting a little longer

I don't think it's worth waiting. Other architectures will have to implement these only once they upgrade `svd2rust` anyways. I have opened issues for the affected architectures to give them a heads up though.

**Comment by adamgreig** [comment]
Oh, thanks, I didn't see the edit to the first comment. SGTM to me then.

**Review by Emilgardis — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+- Use `critical_section::with` instead of `interrupt::free` for `Peripherals::take`.
```
Comment (Emilgardis) [inline CHANGELOG.md:10]: A mention to `take` not being provided if `critical-section` is not enabled could probably be added here

**Review by therealprof — APPROVED** [review]
LGTM,too. Thanks a lot @reitermarkus for taking this on.

bors r+

**Pull request merged**